src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs

```diff
@@ -888,11 +888,32 @@ private void RetrieveBoundOperationPaths(OpenApiConvertSettings convertSettings)
                     continue;
                 }
 
+                ODataSegment secondLastSeg = functionPath.ElementAt(functionPath.Count - 2);
+                if (convertSettings.ComposableFunctionsExpansionDepth < 2 &&
+                            functionPath.LastSegment is ODataOperationSegment &&
+                            secondLastSeg is ODataOperationSegment)
+                {
+                    // Only one level of composable functions expansion allowed
+                    continue;
+                }
+
                 foreach (var navProperty in returnBindingEntityType.NavigationProperties())
                 {
-                    ODataPath newNavigationPath = functionPath.Clone();
-                    newNavigationPath.Push(new ODataNavigationPropertySegment(navProperty));
-                    AppendPath(newNavigationPath);
+                    /* Get number of segments already appended after the first composable function segment
+                     */
+                    int composableFuncSegIndex = functionPath.Segments.IndexOf(
+                        functionPath.Segments.FirstOrDefault(
+                            x => x is ODataOperationSegment operationSegment &&
+                            operationSegment.Operation is IEdmFunction edmFunction &&
+                            edmFunction.IsComposable));                   
+                    int currentDepth = functionPath.Count - composableFuncSegIndex - 1;
+
+                    if (currentDepth < convertSettings.ComposableFunctionsExpansionDepth)
```

Comment:
```suggestion
                    if (currentDepth < convertSettings.ComposableFunctionsExpansionDepth && composableFuncSegIndex >= 0)
```
We probably want to be sure that the search using `IndexOf` was successful in line 904

Reply:
Since at that point we are sure we have at least 1 operation segment that `IsComposable = true`, `composableFuncSegIndex` will never be a negative integer. However, to fool proof this, we can consider putting the safeguard when initializing the `currentDepth` variable:
`int currentDepth = (composableFuncSegIndex > 0) ?  functionPath.Count - composableFuncSegIndex - 1 : 0;`

Reply:
Thanks for clarifying this. After taking further look, looks like we have the guarantee that the last segment will at least meet this condition. Resolving this for now ....